File: document/src/main/java/com/yahoo/document/fieldset/FieldSetRepo.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.document.fieldset;

import com.yahoo.document.Document;
import com.yahoo.document.DocumentType;
import com.yahoo.document.DocumentTypeManager;
import com.yahoo.document.Field;
import com.yahoo.document.datatypes.FieldValue;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.StringTokenizer;

/**
 * TODO: Move to document and implement
 */
public class FieldSetRepo {

    FieldSet parseSpecialValues(String name)
    {
        if (name.equals(DocIdOnly.NAME)) { return new DocIdOnly(); }
        else if (name.equals(DocumentOnly.NAME)) { return (new DocumentOnly()); }
        else if (name.equals(AllFields.NAME)) { return (new AllFields()); }
        else if (name.equals(NoFields.NAME)) { return (new NoFields()); }
        else if (name.equals("[docid]")) { return (new DocIdOnly()); }
        else {
            throw new IllegalArgumentException(
                    "The only special names (enclosed in '[]') allowed are " +
                    "id, all, document, none");
        }
    }

    FieldSet parseFieldCollection(DocumentTypeManager docMan, String docType, String fieldNames) {
        DocumentType type = docMan.getDocumentType(docType);
        if (type == null) {
            throw new IllegalArgumentException("Unknown document type " + docType);
        }

        FieldCollection collection = new FieldCollection(type);
        if (fieldNames.equals("[document]")) {
            collection.addAll(type.fieldSet());
        }
        else {
            StringTokenizer tokenizer = new StringTokenizer(fieldNames, ",");
            while (tokenizer.hasMoreTokens()) {
                String token = tokenizer.nextToken();
                Field f = type.getField(token);
                if (f == null) {
                    throw new IllegalArgumentException("No such field " + token);
                }
                collection.add(f);
            }
        }
        return collection;
    }

    public FieldSet parse(DocumentTypeManager docMan, String fieldSet) {
        if (fieldSet.length() == 0) {
            throw new IllegalArgumentException("Illegal field set value \"\"");
        }

        if (fieldSet.startsWith("[")) {
            return parseSpecialValues(fieldSet);
        }

        StringTokenizer tokenizer = new StringTokenizer(fieldSet, ":");
        if (tokenizer.countTokens() != 2) {
            throw new IllegalArgumentException(
                    "The field set list must consist of a document type, " +
                    "then a colon (:), then a comma-separated list of field names");
        }

        String type = tokenizer.nextToken();
        String fields = tokenizer.nextToken();

        return parseFieldCollection(docMan, type, fields);
    }

    public String serialize(FieldSet fieldSet) {
        if (fieldSet instanceof Field) {
            return ((Field)fieldSet).getName();
        } else if (fieldSet instanceof FieldCollection) {
            FieldCollection c = ((FieldCollection)fieldSet);

            StringBuilder buffer = new StringBuilder();
            for (Field f : c) {
                if (buffer.length() == 0) {
                    buffer.append(c.getDocumentType().getName());
                    buffer.append(":");
                } else {
                    buffer.append(",");
                }
                buffer.append(f.getName());
            }

            return buffer.toString();
        } else if (fieldSet instanceof AllFields) {
            return AllFields.NAME;
        } else if (fieldSet instanceof NoFields) {
            return NoFields.NAME;
        } else if (fieldSet instanceof DocIdOnly) {
            return DocIdOnly.NAME;
        } else if (fieldSet instanceof DocumentOnly) {
            return DocumentOnly.NAME;
        } else {
            throw new IllegalArgumentException("Unknown field set type " + fieldSet);
        }
    }


    /**
     * Copies fields from one document to another based on whether the fields match the given
     * fieldset.
     */
    public void copyFields(Document source, Document target, FieldSet fieldSet) {
        if (fieldSet instanceof DocumentOnly) {
            var actual = source.getDataType().fieldSet(DocumentOnly.NAME);
            if (actual != null) {
                for (Iterator<Map.Entry<Field, FieldValue>> i = source.iterator(); i.hasNext();) {
                    Map.Entry<Field, FieldValue> v = i.next();
                    if (actual.contains(v.getKey())) {
                        target.setFieldValue(v.getKey(), v.getValue());
                    }
                }
                return;
            }
        }
        for (Iterator<Map.Entry<Field, FieldValue>> i = source.iterator(); i.hasNext();) {
            Map.Entry<Field, FieldValue> v = i.next();

            if (fieldSet.contains(v.getKey())) {
                target.setFieldValue(v.getKey(), v.getValue());
            }
        }
    }

    /**
     * Strips all fields not wanted by the given field set from the document.
     */
    public void stripFields(Document target, FieldSet fieldSet) {
        List<Field> toStrip = new ArrayList<>();
        if (fieldSet instanceof DocumentOnly) {
            var actual = target.getDataType().fieldSet(DocumentOnly.NAME);
            if (actual != null) {
                for (Iterator<Map.Entry<Field, FieldValue>> i = target.iterator(); i.hasNext();) {
                    Map.Entry<Field, FieldValue> v = i.next();
                    if (! actual.contains(v.getKey())) {
                        toStrip.add(v.getKey());
                    }
                }
                for (Field f : toStrip) {
                    target.removeFieldValue(f);
                }
                return;
            }
        }
        for (Iterator<Map.Entry<Field, FieldValue>> i = target.iterator(); i.hasNext();) {
            Map.Entry<Field, FieldValue> v = i.next();

            if (!fieldSet.contains(v.getKey())) {
                toStrip.add(v.getKey());
            }
        }

        for (Field f : toStrip) {
            target.removeFieldValue(f);
        }
    }
}

File: document/src/vespa/document/fieldset/fieldsetrepo.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "fieldsetrepo.h"
#include <vespa/vespalib/text/stringtokenizer.h>
#include <vespa/vespalib/util/exceptions.h>
#include <vespa/vespalib/stllike/asciistream.h>
#include <vespa/document/repo/documenttyperepo.h>
#include <vespa/document/base/exceptions.h>
#include <vespa/vespalib/stllike/hash_map.hpp>

#include <vespa/log/log.h>
LOG_SETUP(".document.fieldset.fieldsetrepo");

using vespalib::StringTokenizer;

namespace document {

namespace {

FieldSet::SP
parseSpecialValues(vespalib::stringref name)
{
    if ((name.size() == 4) && (name[1] == 'i') && (name[2] == 'd') && (name[3] == ']')) {
        return std::make_shared<DocIdOnly>();
    } else if ((name.size() == 5) && (name[1] == 'a') && (name[2] == 'l') && (name[3] == 'l') && (name[4] == ']')) {
        return std::make_shared<AllFields>();
    } else if ((name.size() == 6) && (name[1] == 'n') && (name[2] == 'o') && (name[3] == 'n') && (name[4] == 'e') && (name[5] == ']')) {
        return std::make_shared<NoFields>();
    } else if ((name.size() == 7) && (name[1] == 'd') && (name[2] == 'o') && (name[3] == 'c') && (name[4] == 'i') && (name[5] == 'd') && (name[6] == ']')) {
        return std::make_shared<DocIdOnly>();
    } else {
        throw vespalib::IllegalArgumentException(
                "The only special names (enclosed in '[]') allowed are "
                "id, all, none, not '" + name + "'.");
    }
}

FieldSet::SP
parseFieldCollection(const DocumentTypeRepo& repo,
                     vespalib::stringref docType,
                     vespalib::stringref fieldNames)
{
    const DocumentType* typePtr = repo.getDocumentType(docType);
    if (!typePtr) {
        throw vespalib::IllegalArgumentException("Unknown document type " + docType);
    }
    const DocumentType& type(*typePtr);

    StringTokenizer tokenizer(fieldNames, ",");
    Field::Set::Builder builder;
    for (const auto & token : tokenizer) {
        const DocumentType::FieldSet * fs = type.getFieldSet(token);
        if (fs) {
            for (const auto & fieldName : fs->getFields()) {
                builder.add(&type.getField(fieldName));
            }
        } else {
            builder.add(&type.getField(token));
        }
    }
    return std::make_shared<FieldCollection>(type, builder.build());
}

}

FieldSet::SP
FieldSetRepo::parse(const DocumentTypeRepo& repo, vespalib::stringref str)
{
    if (str[0] == '[') {
        return parseSpecialValues(str);
    } else {
        StringTokenizer tokenizer(str, ":");
        if (tokenizer.size() != 2) {
            throw vespalib::IllegalArgumentException(
                    "The field set list must consist of a document type, "
                    "then a colon (:), then a comma-separated list of field names");
        }

        return parseFieldCollection(repo, tokenizer[0], tokenizer[1]);
    }
}

vespalib::string
FieldSetRepo::serialize(const FieldSet& fieldSet)
{
    switch (fieldSet.getType()) {
        case FieldSet::Type::FIELD:
            return static_cast<const Field&>(fieldSet).getName();
        case FieldSet::Type::SET: {
            const auto & collection = static_cast<const FieldCollection&>(fieldSet);

            vespalib::asciistream stream;
            stream << collection.getDocumentType().getName() << ":";
            bool first = true;
            for (const Field * field : collection.getFields()) {
                if (first) {
                    first = false;
                } else {
                    stream << ",";
                }
                stream << field->getName();
            }

            return stream.str();
        }
        case FieldSet::Type::ALL:
            return AllFields::NAME;
        case FieldSet::Type::NONE:
            return NoFields::NAME;
        case FieldSet::Type::DOCID:
            return DocIdOnly::NAME;
        default:
            return "";
    }
}


FieldSetRepo::FieldSetRepo(const DocumentTypeRepo& repo)
    : _doumentTyperepo(repo),
      _configuredFieldSets()
{
    repo.forEachDocumentType(*DocumentTypeRepo::makeLambda([&](const DocumentType &type) {
        configureDocumentType(type);
    }));
}
FieldSetRepo::~FieldSetRepo() = default;

void
FieldSetRepo::configureDocumentType(const DocumentType & documentType) {
    for (const auto & entry : documentType.getFieldSets()) {
        vespalib::string fieldSetName(documentType.getName());
        fieldSetName.append(':').append(entry.first);
        try {
            auto fieldset = parse(_doumentTyperepo, fieldSetName);
            _configuredFieldSets[fieldSetName] = std::move(fieldset);
        } catch (const FieldNotFoundException & ex) {
            // Just silently skip it so error handling can be done when you can return proper error to user.
        }
    }
}
FieldSet::SP
FieldSetRepo::getFieldSet(vespalib::stringref fieldSetString) const {
    auto found = _configuredFieldSets.find(fieldSetString);
    if (found != _configuredFieldSets.end()) {
        return found->second;
    }
    return parse(_doumentTyperepo, fieldSetString);
}

}

